exec/java-exec/src/main/java/org/apache/drill/exec/store/pcap/PcapRecordReader.java

```diff
@@ -0,0 +1,295 @@
+  }
+
+  @Override
+  public void setup(final OperatorContext context, final OutputMutator output) throws ExecutionSetupException {
+    this.output = output;
+  }
+
+  @Override
+  public int next() {
+    projectedCols = getProjectedColsIfItNull();
+    try {
+      return parsePcapFilesAndPutItToTable();
+    } catch (IOException io) {
+      throw new RuntimeException("Trouble with reading packets in file!");
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+  }
+
+  private PacketDecoder getPacketDecoder() {
+    try {
+      return new PacketDecoder(in);
+    } catch (IOException io) {
+      throw new RuntimeException("File Not Found or some I/O issue");
+    }
+  }
+
+  private ImmutableList<ProjectedColumnInfo> getProjectedColsIfItNull() {
+    return projectedCols != null ? projectedCols : initCols(new Schema());
+  }
+
+  private ImmutableList<ProjectedColumnInfo> initCols(final Schema schema) {
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+    ColumnDto column;
+
+    for (int i = 0; i < schema.getNumberOfColumns(); i++) {
+      column = schema.getColumnByIndex(i);
+
+      final String name = column.getColumnName();
+      final PcapTypes type = column.getColumnType();
+      TypeProtos.MinorType minorType = TYPES.get(type);
+
+      ProjectedColumnInfo pci = getProjectedColumnInfo(column, name, minorType);
+      pciBuilder.add(pci);
+    }
+    return pciBuilder.build();
+  }
+
+  private ProjectedColumnInfo getProjectedColumnInfo(final ColumnDto column,
+                                                     final String name,
+                                                     final MinorType minorType) {
+    TypeProtos.MajorType majorType = getMajorType(minorType);
+
+    MaterializedField field =
+        MaterializedField.create(name, majorType);
+
+    ValueVector vector =
+        getValueVector(minorType, majorType, field);
+
+    return getProjectedColumnInfo(column, vector);
+  }
+
+  private ProjectedColumnInfo getProjectedColumnInfo(final ColumnDto column, final ValueVector vector) {
+    ProjectedColumnInfo pci = new ProjectedColumnInfo();
+    pci.vv = vector;
+    pci.pcapColumn = column;
+    return pci;
+  }
+
+  private MajorType getMajorType(final MinorType minorType) {
+    return Types.optional(minorType);
+  }
+
+  private ValueVector getValueVector(final MinorType minorType,
+                                     final MajorType majorType,
+                                     final MaterializedField field) {
+    try {
+
+      final Class<? extends ValueVector> clazz = TypeHelper.getValueVectorClass(
+          minorType, majorType.getMode());
+      ValueVector vector = output.addField(field, clazz);
+      vector.allocateNew();
+      return vector;
+
+    } catch (SchemaChangeException sce) {
+      throw new NullPointerException("The addition of this field is incompatible with this OutputMutator's capabilities");
```

Comment:
`NullPointerException` --> `IllegalStateException`

This is not really a null pointer, but it is an illegal state...